src/trie/trie.c

```diff
@@ -604,3 +607,163 @@ TrieNode *TrieNode_RandomWalk(TrieNode *n, int minSteps, rune **str, t_len *len)
+  uint16_t n;
+} rsbHelper;
+
+static int rsbCompareCommon(const void *h, const void *e, int onlyCommon) {
+  const rsbHelper *term = h;
+  const TrieNode *elem = *(const TrieNode **)e;
+  size_t ntmp;
+  int rc;
+  if (onlyCommon) {
+    size_t cmplen = MIN(term->n, elem->len);
+    rc = runecmp(term->r, cmplen, elem->str, cmplen);
+  } else {
+    rc = runecmp(term->r, term->n, elem->str, elem->len);
+  }
+  // printf("Comparing h(%s) to e(%s) => %d\n", runesToStr(term->r, term->n, &ntmp),
+  //        runesToStr(elem->str, elem->len, &ntmp), rc);
+  return rc;
+}
+
+static int rsbCompareEnd(const void *h, const void *e) {
+  return rsbCompareCommon(h, e, 0);
+}
+
+static int rsbCompareBegin(const void *h, const void *e) {
+  return rsbCompareCommon(h, e, 1);
+}
+
+typedef struct {
+  rune buf[TRIE_MAX_STRING_LEN + 1];
+  size_t offset;
+  TrieRangeCallback *callback;
+  void *cbctx;
+} RangeCtx;
+
+/**
+ * Try to place as many of the common arguments in rangectx, so that the stack
+ * size is not negatively impacted and prone to attack.
+ */
+static void rangeIterate(TrieNode *n, const rune *min, uint16_t nmin, const rune *max,
+                         uint16_t nmax, RangeCtx *r) {
+  // printf("Enter\n");
+  // Push string to stack
+  memcpy(r->buf + r->offset, n->str, sizeof(*n->str) * n->len);
+  r->offset += n->len;
+  r->buf[r->offset] = 0;
+
+  /**
+   * Chop off the leading chars of the term which are currently unused. If no
+   * chars remain, then remove it
+   */
+  if (nmin < n->len) {
+    nmin = 0;
+    min = NULL;
+  } else {
+    nmin -= n->len;
+    min += n->len;
+  }
+  if (nmax < n->len) {
+    nmax = 0;
+    max = NULL;
+  } else {
+    nmax -= n->len;
+    max += n->len;
+  }
+
+  /**
+   * If this node is terminal AND nmin is 0, invoke the
+   * callback. If nmin is not zero, then it means that this
+   * node is smaller than the minimum (even if it might be a
+   * parent to nodes that do match!)
+   */
+  if (nmin == 0 && __trieNode_isTerminal(n)) {
+    r->callback(r->buf, r->offset, r->cbctx);
+  }
+
+  TrieNode **arr = __trieNode_children(n);
+  size_t arrlen = n->numChildren;
+  if (!arrlen) {
+    // printf("Node has no children..\n");
+    goto clean_stack;
+  }
+
+  if (n->sortmode != TRIENODE_SORTED_LEX) {
+    qsort(arr, arrlen, sizeof(*arr), cmpLexFull);
+    n->sortmode = TRIENODE_SORTED_LEX;
+  } else {
+    // printf("Already sorted lexically wtf!\n");
+  }
+
+  // Find the minimum range here..
+  // Use binary search to find the beginning and end ranges:
+  rsbHelper h;
+
+  uint16_t beginIdx = 0, endIdx = arrlen;
+  if (nmin) {
+    h.r = min;
+    h.n = nmin;
+    beginIdx = rsb_ge(arr, arrlen, sizeof(*arr), 0, arrlen, &h, rsbCompareBegin);
+    if (beginIdx == arrlen) {
+      // printf("First and last element are same. Bad!\n");
+      goto clean_stack;
+    }
+  }
+  if (nmax) {
+    h.r = max;
+    h.n = nmax;
+    endIdx = rsb_ge(arr, arrlen, sizeof(*arr), beginIdx, arrlen, &h, rsbCompareEnd);
```

Comment:
shouldn't it be endindex  instead of beginIdx?

Reply:
no, because we know that the match can't be in any index below `beginIdx` (since `max` is greater than `min`, and since `beginIdx == min`)

Reply:
I see, thought it meant something else, ok looks good.